### PR TITLE
Do full recompilation on fatal failure for Groovy

### PR DESCRIPTION
**Pull request by asodja**
Relates to https://github.com/gradle/gradle/issues/22814

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@octylFractal Is it possible to run GroovyCompile with Java6? I avoided using `Streams` in the solution, since I am not sure if we can still run with Java6 or we only support Java8+

**Comment by asodja** [comment]
@bot-gradle clean

File: subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java [hunk 1]
```diff
@@ -297,6 +306,17 @@ private void copyJavaCompilerResult(ApiCompilerResult javaCompilerResult) {
         }
     }
 
+    private static boolean isFatalException(org.codehaus.groovy.control.CompilationFailedException e) {
+        if (e instanceof org.codehaus.groovy.control.MultipleCompilationErrorsException) {
+            for (Message message : ((MultipleCompilationErrorsException) e).getErrorCollector().getErrors()) {
+                if (message instanceof ExceptionMessage) {
```
Comment (leonard84) [inline subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java:312]: To me it is not immediately obvious why this makes it fatal, a comment would be nice.

**Review by leonard84 — APPROVED** [review]
LGTM as far as I can tell

**Comment by octylFractal** [comment]
> Is it possible to run GroovyCompile with Java6? 

AFAIK not intentionally, in this case `ApiGroovyCompiler` is already compiled to Java 8 bytecode so there is no issues with that specific class.

File: subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java (hunk 1 above)
Comment (lptr) [reply subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java:312]: Maybe include an example exception stacktrace as well?

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractJavaGroovyIncrementalCompilationSupport.groovy [hunk 2]
```diff
@@ -53,7 +53,7 @@ abstract class AbstractJavaGroovyIncrementalCompilationSupport extends AbstractI
             def packageGroup = (body =~ "\\s*package ([\\w.]+).*")
             String packageName = packageGroup.size() > 0 ? packageGroup[0][1] : ""
             String packageFolder = packageName.replaceAll("[.]", "/")
-            def className = (body =~ /(?s).*?(?:class|interface|enum) ([\w$]+) .*/)[0][1]
+            def className = (body =~ /(?s).*?(?:class|interface|enum|trait) ([\w$]+) .*/)[0][1]
```
Comment (lptr) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractJavaGroovyIncrementalCompilationSupport.groovy:56]: ❓ How is this change related to this PR?

Comment (asodja) [reply subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractJavaGroovyIncrementalCompilationSupport.groovy:56]: This code creates a file directly from string source.

In the test we added a source:
```
trait WithTrait {
...
}
```
If we want that `WithTrait.groovy` is correctly created, we also have to parse `className` where class starts with `trait <ClassName>`.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test and merge

**Comment by asodja** [comment]
@bot-gradle cancel clean

**Comment by asodja** [comment]
@bot-gradle cancel

**Comment by asodja** [comment]
@bot-gradle test and merge
